src/JSExpr.jl

```diff
@@ -1,207 +1,91 @@
+    deparse(jsnode)
+    deparse(head, args...)
 
-function func_expr(args, body)
-    parts = []
-    named = isexpr(args, :call)
-    named || push!(parts, "(")
-    push!(parts, "function ")
-    if named
-        push!(parts, string(args.args[1]))
-        args.args = args.args[2:end]
-    end
-    push!(parts, "(")
-    isexpr(args, Symbol) ? push!(parts, string(args)) : push!(parts, jsexpr_joined(args.args, ","))
-    push!(parts, "){")
-    push!(parts, jsexpr(insert_return(body)))
-    push!(parts, "}")
-    named || push!(parts, ")")
-    F(parts)
-end
+The expectation is that each dispatched deparse function returns a bare JSString
+literal, formed by appropriate ordering and concatenation of the output of
+recursive calls to the deparse-function.
 
-function insert_return(ex)
-    if isa(ex, Symbol) || !isexpr(ex, :block)
-        Expr(:return, ex)
-    else
-        isexpr(ex.args[end], :return) && return ex
-        ex1 = copy(ex)
-        ex1.args[end] = insert_return(ex.args[end])
-        ex1
-    end
+Convert a `JSNode` to `JSString`.
+"""
+function deparse(ex::JSAST)::JSString
+    deparse(Val(ex.head), ex.args)::JSString
 end
 
+include("./literals.jl")
+include("./infix.jl")
 
-function dict_expr(xs)
-    parts = []
-    xs = map(xs) do x
-        if x.head == :(=) || x.head == :kw
-            push!(parts, F([jsexpr(x.args[1]), ":", jsexpr(x.args[2])]))
-        elseif x.head == :call && x.args[1] == :(=>)
-            push!(parts, F([jsexpr(x.args[2]), ":", jsexpr(x.args[3])]))
-        else
-            error("Invalid pair separator in dict expression")
-        end
-    end
-    F(["{", F(intersperse(parts, ",")), "}"])
+# Dumps the raw JS string literal into the argument of the parent
+# expression that the macro was called from. The common use
+# cases are for the macro to be called as the RHS of an assignment,
+# or an interpolation into another literal.
+macro js(ex)
+    return Expr(:call, :JSString, deparse(crawl(ex)))
 end
 
-function vect_expr(xs)
-    F(["[", F(intersperse([jsexpr(x) for x in xs], ",")), "]"])
-end
-
-function if_block(ex)
+# All other terminals
+function crawl(ex::T)::JSNode where {T}
 
-    if isexpr(ex, :block)
-        if any(x -> isexpr(x, :macrocall) && x.args[1] == Symbol("@var"), ex.args)
-            error("@js expression error: @var inside an if statement is not supported")
-        end
-        F(["(", block_expr(rmlines(ex).args, ", "), ")"])
-    else
-        jsexpr(ex)
-    end
-end
+    # Push terminals that have native symbol methods, maybe unsafe
+    if hasmethod(JSSymbol, Tuple{T})
+        return JSSymbol(ex)
 
-function if_expr(xs)
-    parts = []
-    if length(xs) >= 2    # we have an if
-        append!(parts, [jsexpr(xs[1]), " ? ", if_block(xs[2])])
-    end
+    # Push terminals, when all else fails try interpolation. This will be a source of bugs
+    # because there are no guarantees the string representation will be sensible.
+    elseif hasmethod(string, Tuple{T})
```

Comment:
This is always true. As above, it can probably be simplified by adding a fallback ``JSSymbol(x::Any) = JSSymbol(JSString(string(ex)))` and calling `JSSymbol(ex)` directly here rather than the `if hasmethod` approach.